### PR TITLE
Add System.Xml ref assembly

### PR DESCRIPTION
**Pull request by sepidehMS**
Addresses issue https://github.com/dotnet/corefx/issues/11888

All new System.Xml and System.Xml.* APIs are now added to System.Xml.ReaderWriter ref assembly. All other Xml contracts have also been updated.

cc: @danmosemsft 
cc: @ericstj @weshaggard @karajas  https://github.com/dotnet/corefx/commit/3a9782afdf90c457155c2513321831578a8d6443 contains changes to fix the default TestTFM conditions in `dir.props`

### CODE REVIEW

File: src/System.Private.Xml/src/System.Private.Xml.csproj [hunk 1]
```diff
@@ -817,13 +821,13 @@
     <Compile Include="Workarounds\DummyTypes\PermissionSet.cs" />
     <Compile Include="Workarounds\DummyTypes\RequestCachePolicy.cs" />
     <Compile Include="Workarounds\DummyTypes\SecurityManger.cs" />
-    <Compile Include="Workarounds\SerializationEntry.cs" />
-    <Compile Include="Workarounds\ICloneable.cs" />
+    <None Include="Workarounds\SerializationEntry.cs" />
```
Comment (danmosemsft) [inline src/System.Private.Xml/src/System.Private.Xml.csproj]: >Include [](start = 10, length = 7)

if we don't need these workarounds, we should delete the source files entirely shouldn't we?

File: src/System.Private.Xml/src/System/Xml/Serialization/Compilation.cs [hunk 2]
```diff
@@ -169,8 +169,8 @@ internal static Assembly LoadGeneratedAssembly(Type type, string defaultNamespac
             }*/
 
             // check to see if we loading explicit pre-generated assembly
-            IEnumerable<Attribute> attrs = type.GetTypeInfo().GetCustomAttributes(typeof(XmlSerializerAssemblyAttribute), false);
-            if (attrs.Count() == 0)
+            object[] attrs = type.GetCustomAttributes(typeof(XmlSerializerAssemblyAttribute), false);
```
Comment (danmosemsft) [inline src/System.Private.Xml/src/System/Xml/Serialization/Compilation.cs]: >object [](start = 12, length = 6)

this is all just cleanup because GetTypeInfo is no longer necesasry?

File: src/System.Xml.XmlSerializer/src/project.json [hunk 3]
```diff
@@ -32,10 +32,10 @@
         "System.Threading.Tasks": "4.3.0-beta-24522-03"
       },
       "imports": [
```
Comment (danmosemsft) [inline src/System.Xml.XmlSerializer/src/project.json]: >imports [](start = 7, length = 7)

I thought we didn't need imports, but I don't know much about this.

**Comment by danmosemsft** [comment]
I think you said you were going to do a diff (in API reviewer or whatever) between your new XML reference assemblies and the netstandard.dll built out of Wes's repo, right? There should be zero differences -- any differences should have an issue assigned to you for them.

File: src/System.Private.Xml/src/System/Xml/Serialization/Compilation.cs [hunk 4]
```diff
@@ -169,8 +169,8 @@ internal static Assembly LoadGeneratedAssembly(Type type, string defaultNamespac
...
+            if (attrs.Length == 0)
             {
                 // Guess serializer name: if parent assembly signed use strong name 
                 AssemblyName name = type.GetTypeInfo().Assembly.GetName();
```
Comment (danmosemsft) [inline src/System.Private.Xml/src/System/Xml/Serialization/Compilation.cs]: >type.GetTypeInfo() [](start = 36, length = 18)

not suggesting cleaning these up now, probably not worth even a tracking issue, but according to code flow there are still 64 cases where the code does type.GEtTypeInfo().

File: src/System.Private.Xml/src/System/Xml/Xsl/QIL/QilXmlReader.cs [hunk 5]
```diff
@@ -368,6 +368,7 @@ private void EndElement()
                         MethodInfo mi = null;
                         QilName name = (QilName)list[0];
 
+#pragma warning disable 0436  //TODO: remove this when AppDomain.GetAssemblies() becomes available
```
Comment (danmosemsft) [inline src/System.Private.Xml/src/System/Xml/Xsl/QIL/QilXmlReader.cs]: >TODO [](start = 32, length = 4)

issue needed

Comment (danmosemsft) [inline src/System.Private.Xml/src/System/Xml/Xsl/QIL/QilXmlReader.cs]: actually, it's already ocvered by your issue to remove the workaroudn stubs.

---
In reply to: [80328855](https://github.com/dotnet/corefx/pull/12017#discussion_r80328855) [](ancestors = 80328855)

File: src/System.Private.Xml.Linq/src/System/Xml/Linq/XName.cs [hunk 6]
```diff
@@ -181,5 +182,38 @@ bool IEquatable<XName>.Equals(XName other)
...
             return (object)this == (object)other;
         }
+
+        /// <summary>
+        /// Populates a <see cref="SerializationInfo"/> with the data needed to
+        /// serialize the <see cref="XName"/>
+        /// </summary>
+        /// <param name="info">The <see cref="SerializationInfo"/> to populate with data</param>
+        /// <param name="context">The destination for this serialization</param>
+        void ISerializable.GetObjectData(SerializationInfo info, StreamingContext context) {
```
Comment (danmosemsft) [inline src/System.Private.Xml.Linq/src/System/Xml/Linq/XName.cs]: >ISerializable [](start = 13, length = 13)

do we need tests for all these serializable implementations you're adding?

**Comment by danmosemsft** [comment]
Please get signoff from @shmao or someone on his team as this touches XML Serializer.

**Comment by shmao** [comment]
Thanks for the heads up, @danmosemsft. 

cc: @zhenlan @mconnew @huanwu

File: src/System.Private.Xml/src/System/Xml/Serialization/Compilation.cs (hunk 2 above)
Comment (sepidehMS) [inline src/System.Private.Xml/src/System/Xml/Serialization/Compilation.cs]: This code wouldn't build anymore as `MemberInfo.GetCustomAttributes` now returns `object[]` which is cannot implicitly be converted to `IEnumerable<>`. We used `GetTypeInfo` as a workaround when porting Xml, now I changed them back to what they used to be.

File: src/System.Private.Xml/src/System.Private.Xml.csproj (hunk 1 above)
Comment (sepidehMS) [inline src/System.Private.Xml/src/System.Private.Xml.csproj]: Will do. (**Note:** I'm going to do a cleanup for all the XML workarounds soon tracked by issue https://github.com/dotnet/corefx/issues/12000)

File: src/System.Private.Xml/src/System/Xml/Serialization/Compilation.cs (hunk 4 above)
Comment (sepidehMS) [inline src/System.Private.Xml/src/System/Xml/Serialization/Compilation.cs]: Same as above for `GetTypeInfo()`

File: src/System.Xml.XmlSerializer/src/project.json (hunk 3 above)
Comment (weshaggard) [inline src/System.Xml.XmlSerializer/src/project.json]: They can be removed.

File: dir.props [hunk 7]
```diff
@@ -487,18 +487,10 @@
 
   <!-- Default Test platform to deploy the netstandard compiled tests to -->
   <PropertyGroup>
-    <TestTFM Condition="'$(TestTFM)'=='' and ('$(TargetGroup)'=='netstandard1.7' or '$(TargetGroup)'=='netcoreapp1.1')">netcoreapp1.1</TestTFM>
```
Comment (weshaggard) [inline dir.props]: This looks unrelated to the XML changes. Is there a reason you needed to move these to dir.targets as part of your work?

File: src/System.Private.Xml.Linq/src/System.Private.Xml.Linq.csproj [hunk 8]
```diff
@@ -9,15 +9,15 @@
     <DefineConstants>$(DefineConstants);SILVERLIGHT</DefineConstants>
     <IsPartialFacadeAssembly Condition="'$(TargetGroup)' == 'net46'">true</IsPartialFacadeAssembly>
     <ResourcesSourceOutputDirectory Condition="'$(TargetGroup)' == 'net46'">None</ResourcesSourceOutputDirectory>
```
Comment (weshaggard) [inline src/System.Private.Xml.Linq/src/System.Private.Xml.Linq.csproj]: Orthogonal to your change but do you know if there is any reason this library cannot be one-to-one with the System.Xml.XDocument contract? I didn't believe it had any circular dependencies as it was already implemented independently on the full .NET framework.

File: src/System.Private.Xml.Linq/src/System/Xml/Linq/XName.cs [hunk 9]
```diff
@@ -181,5 +182,38 @@ bool IEquatable<XName>.Equals(XName other)
...
+            if (info == null) throw new ArgumentNullException("info");
+            info.AddValue("name", ToString());
+            info.SetType(typeof(NameSerializer));
+        }
+    }
+
+#if SERIALIZABLE_DEFINED
```
Comment (weshaggard) [inline src/System.Private.Xml.Linq/src/System/Xml/Linq/XName.cs]: I assume this #ifdef isn't needed.

File: src/System.Private.Xml.Linq/src/project.json [hunk 10]
```diff
@@ -22,14 +22,15 @@
         "System.Reflection": "4.3.0-beta-24522-03",
         "System.Runtime.Extensions": "4.3.0-beta-24522-03",
         "System.Threading": "4.3.0-beta-24522-03",
+        "System.Threading.Tasks": "4.3.0-beta-24522-03",
         "System.Threading.Tasks.Extensions": "4.3.0-beta-24522-03",
         "System.Text.Encoding": "4.3.0-beta-24522-03"
       },
       "imports": [
```
Comment (weshaggard) [inline src/System.Private.Xml.Linq/src/project.json]: These imports are no longer needed.

File: src/System.Private.Xml/src/System.Private.Xml.builds [hunk 11]
```diff
@@ -3,12 +3,12 @@
   <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.props))\dir.props" />
   <ItemGroup>
     <Project Include="System.Private.Xml.csproj" />
-    <Project Include="System.Private.Xml.csproj">
+    <!--<Project Include="System.Private.Xml.csproj">
```
Comment (weshaggard) [inline src/System.Private.Xml/src/System.Private.Xml.builds]: We do we not need these builds any longer? If we don't actually need them then we should remove them instead of commenting out. If we have to disable them temporarily then add a comment with a link to an issue to fix them.

File: src/System.Private.Xml/src/System.Private.Xml.csproj [hunk 12]
```diff
@@ -9,16 +9,20 @@
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
     <NoWarn>$(NoWarn),649</NoWarn>
     <DefineConstants>ASYNC;ASTORIA_LIGHT</DefineConstants><!-- TODO: this should be prepended with $(DefineConstants); - otherwise DEBUG and TRACE are not being passed -->
-    <DefineConstants Condition="'$(TargetGroup)' == 'netcore50aot'">$(DefineConstants);NET_NATIVE</DefineConstants>
-    <PackageTargetFramework Condition="'$(PackageTargetFramework)' == ''">netstandard1.3</PackageTargetFramework>
-    <NuGetTargetMoniker Condition="'$(TargetGroup)' == ''">.NETStandard,Version=v1.3</NuGetTargetMoniker>
+    <DefineConstants Condition="'$(TargetGroup)' == 'uap101aot'">$(DefineConstants);NET_NATIVE</DefineConstants>
```
Comment (weshaggard) [inline src/System.Private.Xml/src/System.Private.Xml.csproj]: I don't see any uap101aot build configurations in the .builds file.

File: src/System.Private.Xml/src/System.Private.Xml.csproj [hunk 13]
```diff
@@ -9,16 +9,20 @@
...
+    <PackageTargetFramework Condition="'$(PackageTargetFramework)' == ''">netstandard1.7</PackageTargetFramework>
```
Comment (weshaggard) [inline src/System.Private.Xml/src/System.Private.Xml.csproj]: PackageTargetFramework property isn't needed as it is redundant with NugetTargetMoniker.

File: src/System.Private.Xml/src/System.Private.Xml.csproj [hunk 14]
```diff
@@ -9,16 +9,20 @@
...
   <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Debug|AnyCPU' " />
   <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Release|AnyCPU' " />
-  <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'net46_Debug|AnyCPU'" />
-  <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'net46_Release|AnyCPU'" />
-  <ItemGroup Condition="'$(TargetGroup)' != 'net46'">
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'net463_Debug|AnyCPU'" />
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'net463_Release|AnyCPU'" />
+  <ItemGroup>
+    <!-- remove this project reference when new packages are generated -->
+    <ProjectReference Include="..\..\System.IO.FileSystem\ref\System.IO.FileSystem.csproj" />
```
Comment (weshaggard) [inline src/System.Private.Xml/src/System.Private.Xml.csproj]: What are you depending on that needs the live project reference?

File: src/System.Private.Xml/src/System/Xml/Schema/XmlSchemaException.cs [hunk 15]
```diff
@@ -16,7 +16,7 @@ namespace System.Xml.Schema
 #if SERIALIZABLE_DEFINED
```
Comment (weshaggard) [inline src/System.Private.Xml/src/System/Xml/Schema/XmlSchemaException.cs]: #ifdef shouldn't be needed. You should be able to clean out all those defines.

File: src/System.Private.Xml/src/System/Xml/Serialization/SoapAttributes.cs [hunk 16]
```diff
@@ -43,38 +43,31 @@ public SoapAttributes()
...
-        public SoapAttributes(MemberInfo memberInfo)
+        public SoapAttributes(ICustomAttributeProvider provider)
         {
-            foreach (Attribute attr in memberInfo.GetCustomAttributes(false))
-            {
-                if (attr is SoapIgnoreAttribute || attr is ObsoleteAttribute)
-                {
+            object[] attrs = provider.GetCustomAttributes(false);
+            for (int i = 0; i < attrs.Length; i++) {
+                if (attrs[i] is SoapIgnoreAttribute || attrs[i] is ObsoleteAttribute) {
```
Comment (weshaggard) [inline src/System.Private.Xml/src/System/Xml/Serialization/SoapAttributes.cs]: put braces on a new line.

File: src/System.Private.Xml/tests/XmlSerializer/System.Xml.XmlSerializer.Tests.csproj [hunk 17]
```diff
@@ -11,15 +11,9 @@
...
-        Until we get first class support for NS1.7 and Netcoreapp1.1
-        we need to hard code $(TestTFM) and $(TestNugetTargetMoniker)
-        in the project file.
-      -->
-      <NugetTargetMoniker>.NETStandard,Version=v1.7</NugetTargetMoniker>
-      <TestTFM>netcoreapp1.1</TestTFM>
-      <KeepAllProjectReferences>true</KeepAllProjectReferences>
-    </PropertyGroup>
+    <NugetTargetMoniker>.NETStandard,Version=v1.7</NugetTargetMoniker>
+    <KeepAllProjectReferences>true</KeepAllProjectReferences>
```
Comment (weshaggard) [inline src/System.Private.Xml/tests/XmlSerializer/System.Xml.XmlSerializer.Tests.csproj]: We need an issue to remove KeepAllProjectReferences as that ends up opt'ing it out of testing against our packages during our official builds which can mean we have packaging bugs that go unnoticed in our test passes.

File: src/System.Xml.XDocument/src/System.Xml.XDocument.builds [hunk 18]
```diff
@@ -7,6 +7,9 @@
     <Project Include="System.Xml.XDocument.csproj">
```
Comment (weshaggard) [inline src/System.Xml.XDocument/src/System.Xml.XDocument.builds]: You can remove the net46 build configuration that is commented out.

**Comment by weshaggard** [comment]
@sepidehMS did you not need to move any APIs down into System.Xml.ReaderWriter as part of the API bring back? It is hard to tell from the diff on github.

File: src/System.Xml.XmlSerializer/src/project.json (hunk 3 above)
Comment (sepidehMS) [inline src/System.Xml.XmlSerializer/src/project.json]: Will do

File: src/System.Private.Xml.Linq/src/project.json (hunk 10 above)
Comment (sepidehMS) [inline src/System.Private.Xml.Linq/src/project.json]: Will do.

File: src/System.Private.Xml/src/System.Private.Xml.csproj (hunk 13 above)
Comment (sepidehMS) [inline src/System.Private.Xml/src/System.Private.Xml.csproj]: Will do.

File: src/System.Private.Xml/src/System/Xml/Serialization/SoapAttributes.cs (hunk 16 above)
Comment (sepidehMS) [inline src/System.Private.Xml/src/System/Xml/Serialization/SoapAttributes.cs]: Will do.

**Comment by sepidehMS** [comment]
in reply to https://github.com/dotnet/corefx/pull/12017#issuecomment-249350013 @weshaggard I did need to move some APIs down mostly from XPath and XmlDocument

File: src/System.Private.Xml/src/System.Private.Xml.csproj (hunk 12 above)
Comment (sepidehMS) [inline src/System.Private.Xml/src/System.Private.Xml.csproj]: Since I believe we would need to add uap101aot support for these later, I decided to update the netcore50aot constants to uap101aot now that I'm upgrading this to netstandard1.7 other than commenting it out. If that is not a good move, I can undo it.

File: src/System.Private.Xml.Linq/src/System/Xml/Linq/XName.cs (hunk 9 above)
Comment (sepidehMS) [inline src/System.Private.Xml.Linq/src/System/Xml/Linq/XName.cs]: Going to remove all #ifdef s like this.

File: src/System.Private.Xml/src/System/Xml/Schema/XmlSchemaException.cs (hunk 15 above)
Comment (sepidehMS) [inline src/System.Private.Xml/src/System/Xml/Schema/XmlSchemaException.cs]: Going to remove all.

File: src/System.Xml.XDocument/src/System.Xml.XDocument.builds (hunk 18 above)
Comment (sepidehMS) [inline src/System.Xml.XDocument/src/System.Xml.XDocument.builds]: Will do.

File: src/System.Private.Xml/tests/XmlSerializer/System.Xml.XmlSerializer.Tests.csproj (hunk 17 above)
Comment (sepidehMS) [inline src/System.Private.Xml/tests/XmlSerializer/System.Xml.XmlSerializer.Tests.csproj]: Created an issue to myself.

File: dir.props (hunk 7 above)
Comment (sepidehMS) [inline dir.props]: This code was causing some of the Xml tests to fail, as I think it forced them to run against netcoreapp1.0. I had to set $(TestTFM) and $(TestNugetTargetMoniker) in the csproj to workaround this issue. I talked to @ericstj and he helped figure out this is a bug in dir.props and he suggested this fix for it.

File: src/System.Private.Xml.Linq/src/System.Private.Xml.Linq.csproj (hunk 8 above)
Comment (sepidehMS) [inline src/System.Private.Xml.Linq/src/System.Private.Xml.Linq.csproj]: What does one-to-one mean here?

File: src/System.Private.Xml/src/System.Private.Xml.builds (hunk 11 above)
Comment (sepidehMS) [inline src/System.Private.Xml/src/System.Private.Xml.builds]: Please correct me if I'm wrong. I believe we don't need these because all the Xml packages for netcore50 and netcore50aot are not using System.Private.Xml and will come from package harvesting anyway. So, there is no point in building this package for these configurations.

**Comment by sepidehMS** [comment]
@dotnet-bot test Innerloop Ubuntu14.04 Release Build and Test please (build timed out. error MSB4166: Child node "2" exited prematurely.)

**Comment by sepidehMS** [comment]
In reply to https://github.com/dotnet/corefx/pull/12017#issuecomment-249304453, I did the diff and updated the PR with the differences.

File: src/System.Private.Xml.Linq/src/System.Private.Xml.Linq.csproj (hunk 8 above)
Comment (weshaggard) [inline src/System.Private.Xml.Linq/src/System.Private.Xml.Linq.csproj]: one-to-one in this context is that the implementation should be the same identity as the contract. Both hopefully can be System.Xml.XDocument. That can only happen if they have the same public surface area though.

File: src/System.Private.Xml/src/System.Private.Xml.builds (hunk 11 above)
Comment (weshaggard) [inline src/System.Private.Xml/src/System.Private.Xml.builds]: I believe that is true as well, but it should be verified. If it is true then we should delete the build configurations instead of commenting them out.

**Comment by sepidehMS** [comment]
@dotnet-bot test Innerloop Ubuntu14.04 Debug Build and Test please

File: src/System.Private.Xml/src/System.Private.Xml.csproj (hunk 12 above)
Comment (weshaggard) [inline src/System.Private.Xml/src/System.Private.Xml.csproj]: If you want to keep them I would just go ahead and add the uap101aot build configuration so that we know they are still valid otherwise they will just go stale and there isn't much reason to keep them now.

**Comment by sepidehMS** [comment]
My builds are failing with dependency failures because they expect an older version of `System.Net.Request `to be used, but I need the latest version which has [this](https://github.com/dotnet/corefx/commit/b70d792a0e39312fc4e610a350cf6fce80ab12dd) change. Looks like the packages have not been updated in project.json files. @weshaggard do you know when this change will be applied?

File: src/System.Private.Xml/src/System.Private.Xml.csproj (hunk 14 above)
Comment (sepidehMS) [inline src/System.Private.Xml/src/System.Private.Xml.csproj]: Initially, this was added while I was dealing with some failures which I later fixed. Removed this in a later commit.

File: src/System.Private.Xml/src/System.Private.Xml.csproj (hunk 12 above)
Comment (sepidehMS) [inline src/System.Private.Xml/src/System.Private.Xml.csproj]: AFAIK if I x-compile for uap101aot in this project, I might break a future package update, (as its dependencies are not supporting uap yet) like what happened with IsolatedStorage before. For that reason, I'm going to comment this out for now explaining we need to bring it back when we add uap101aot.

**Comment by sepidehMS** [comment]
@shmao @zhenlan @mconnew @huanwu are there any comments you want to add?

File: src/System.Xml.XmlSerializer/src/System.Xml.XmlSerializer.builds [hunk 19]
```diff
@@ -8,8 +8,11 @@
       <TargetGroup>net46</TargetGroup>
     </Project> -->
     <Project Include="System.Xml.XmlSerializer.csproj">
-      <TargetGroup>uap101aot</TargetGroup>
+      <TargetGroup>net463</TargetGroup>
     </Project>
+    <!--<Project Include="System.Xml.XmlSerializer.csproj">
+      <TargetGroup>uap101aot</TargetGroup>
```
Comment (shmao) [inline src/System.Xml.XmlSerializer/src/System.Xml.XmlSerializer.builds]: Nit: can you please add a comment for why we commented out this group? Thanks.

**Comment by shmao** [comment]
:shipit:

**Comment by shmao** [comment]
Changes related to XmlSerializer LGTM. Thanks for making the change, Sepideh!

File: src/System.Xml.XmlSerializer/src/System.Xml.XmlSerializer.builds (hunk 19 above)
Comment (sepidehMS) [inline src/System.Xml.XmlSerializer/src/System.Xml.XmlSerializer.builds]: Will do.

**Comment by sepidehMS** [comment]
@dotnet-bot test Innerloop Windows_NT Debug Build and Test please
@dotnet-bot test Innerloop Windows_NT Release Build and Test please
(error MSB4015: The build stopped unexpectedly because the "ReusableLogger" logger failed unexpectedly during shutdown.)

**Comment by weshaggard** [comment]
>My builds are failing with dependency failures because they expect an older version of System.Net.Requestto be used, but I need the latest version which has this change. Looks like the packages have not been updated in project.json files. @weshaggard do you know when this change will be applied?

The package update is blocked for various reasons right now all being worked on. You can probably get away with doing a pkgproj ProjectReference for now to get what you need.

**Comment by sepidehMS** [comment]
@weshaggard Do you have any additional comments, or can I merge if CI builds pass?

**Comment by weshaggard** [comment]
@sepidehMS have you done an API diff between these and netstandard.dll to ensure everything matches?

Can you please file a few follow-up issues?
1. Contract naming. A lot of APIs are in System.Xml.ReaderWriter is that the right place and name for them?
2. The todo's for project to project reference clean-up
3. The UAP101 build configuration. If we plan to keep those things in and commented out we should have a follow-up issue to fix.


Other then those the change LGTM

**Comment by sepidehMS** [comment]
@weshaggard Yes, I have done the diff and everything matches with standard.dll now.

Create these follow-up issues:
https://github.com/dotnet/corefx/issues/12144
https://github.com/dotnet/corefx/issues/12145
https://github.com/dotnet/corefx/issues/12146

Thanks for your feedback.
